src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPool.IO.Windows.cs

```diff
@@ -0,0 +1,186 @@
+            {
+                EnsureIOCompletionPollers();
+            }
+
+            nint port = Interop.Kernel32.CreateIoCompletionPort(handle, _ioPort, UIntPtr.Zero, 0);
+            if (port == 0)
+            {
+                ThrowHelper.ThrowApplicationException(Marshal.GetHRForLastWin32Error());
+            }
+
+            Debug.Assert(port == _ioPort);
+        }
+
+        public unsafe void QueueNativeOverlapped(NativeOverlapped* nativeOverlapped)
+        {
+            Debug.Assert(nativeOverlapped != null);
+            Debug.Assert(_ioPort != 0);
+
+            if (_ioCompletionPollers == null)
+            {
+                EnsureIOCompletionPollers();
+            }
+
+            if (NativeRuntimeEventSource.Log.IsEnabled())
+            {
+                NativeRuntimeEventSource.Log.ThreadPoolIOEnqueue(nativeOverlapped);
+            }
+
+            if (!Interop.Kernel32.PostQueuedCompletionStatus(_ioPort, 0, UIntPtr.Zero, (IntPtr)nativeOverlapped))
+            {
+                ThrowHelper.ThrowApplicationException(Marshal.GetHRForLastWin32Error());
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private void EnsureIOCompletionPollers()
+        {
+            _threadAdjustmentLock.Acquire();
+            try
+            {
+                if (_ioCompletionPollers != null)
+                {
+                    return;
+                }
+
+                int pollerCount = (Environment.ProcessorCount - 1) / ProcessorsPerPoller + 1;
+                IOCompletionPoller[] pollers = new IOCompletionPoller[pollerCount];
+                for (int i = 0; i < pollerCount; ++i)
+                {
+                    pollers[i] = new IOCompletionPoller(_ioPort);
+                }
+
+                _ioCompletionPollers = pollers;
+            }
+            catch (Exception ex)
+            {
+                Environment.FailFast("Failed to initialize IO completion pollers.", ex);
+            }
+            finally
+            {
+                _threadAdjustmentLock.Release();
+            }
+        }
+
+        private sealed unsafe class IOCompletionPoller
+        {
+            private const int NativeEventCapacity =
+#if DEBUG
+                32;
+#else
+                1024;
+#endif
+
+            private static readonly Action<Event> ProcessEventDelegate = ProcessEvent;
+
+            private readonly nint _port;
+            private readonly Interop.Kernel32.OVERLAPPED_ENTRY* _nativeEvents;
+            private readonly ThreadPoolTypedWorkItemQueue<Event> _events;
+            private readonly Thread _thread;
+
+            public IOCompletionPoller(nint port)
+            {
+                Debug.Assert(port != 0);
+                _port = port;
+
+                _nativeEvents =
+                    (Interop.Kernel32.OVERLAPPED_ENTRY*)
+                    Marshal.AllocHGlobal(NativeEventCapacity * Marshal.SizeOf<Interop.Kernel32.OVERLAPPED_ENTRY>());
```

Comment:
My understanding is that the `NativeMemory` APIs are preferred when allocating memory (correct me if I am wrong), and [`Unsafe.SizeOf` gets optimized](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgBUALWbAEwEsAdgHMA3DXpNWAJQCuAjH3wwWASXkwoEAA4BlDQDc+YGLjG1GzFrPmLlAYQj4tfADYa9UQ8dM0aAbR0MKBkwDAAZbABPCBkMAAoI6NiAaUEeFj0ARxkYG2wXAEoAXRpBDA0BfIY5XGwAMxgGXCCQjAYAeQA1AFEpMIBBAAVB7oARAH1ugDk2KQBNGgBvGgZVpgBmBgBVNQxBoIYXLQcnNwUIAWSYSLM1jYYp7AV9GHaXqBdsLS0YHgAqQ5aN4aT7fX63NbETY7eT7KAMXYVfIQ1ZQ6plBg8ADuUxk+GAGnadQAQpFyrg2FBsAJcA0oLAeCjJEgGBidHwAF6vOoAWTiBQYAF4AHwMHnYKC4Dj5DKc7kAHi6vQGwzGkxm82F/KZzBZbLlRK2/KFoq2NPqynZXKJip6fSGIwm01mcy1BTMAF8aEA=).

```suggestion
                    NativeMemory.Alloc(NativeEventCapacity, Unsafe.SizeOf<Interop.Kernel32.OVERLAPPED_ENTRY>());
```

Reply:
Yea looks like that would be slightly better, fixed in latest commit